MapboxNavigation/NavigationMapView.swift

```diff
@@ -34,75 +35,223 @@ open class NavigationMapView: MGLMapView {
+        if sender.state == .changed {
+            guard let location = userLocationForCourseTracking else { return }
+            userCourseView?.layer.removeAllAnimations()
+            userCourseView?.center = convert(location.coordinate, toPointTo: self)
+        }
+    }
+    
+    var shouldPositionCourseViewFrameByFrame = false
+    
+    // Track position on a frame by frame basis. Used for first location update and when resuming tracking mode
+    func enableFrameByFrameCourseViewTracking(for duration: TimeInterval) {
+        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(disableFrameByFramePositioning), object: nil)
+        perform(#selector(disableFrameByFramePositioning), with: nil, afterDelay: duration)
+        shouldPositionCourseViewFrameByFrame = true
+    }
+    
+    @objc fileprivate func disableFrameByFramePositioning() {
+        shouldPositionCourseViewFrameByFrame = false
+    }
+    
+    func resetInactivityTimer(_ sender: UIGestureRecognizer) {
+        if sender.state == .began {
+            isInactive = false
+        }
+        else if sender.state == .ended || sender.state == .failed {
+            resetInactivityTimer()
+        }
+    }
+    
     var showsRoute: Bool {
         get {
             return style?.layer(withIdentifier: routeLayerIdentifier) != nil
         }
     }
     
-    open override var delegate: MGLMapViewDelegate? {
-        didSet {
-            refreshShowsUserLocation()
+    public weak var navigationMapDelegate: NavigationMapViewDelegate?
+    weak var courseTrackingDelegate: NavigationMapViewCourseTrackingDelegate!
+    
+    open override var showsUserLocation: Bool {
+        get {
+            if tracksUserCourse || userLocationForCourseTracking != nil {
+                return !(userCourseView?.isHidden ?? true)
+            }
+            return super.showsUserLocation
         }
+        set {
+            if tracksUserCourse || userLocationForCourseTracking != nil {
+                super.showsUserLocation = false
+                
+                if userCourseView == nil {
+                    userCourseView = UserPuckCourseView(frame: CGRect(origin: .zero, size: CGSize(width: 75, height: 75)))
+                }
+                userCourseView?.isHidden = !newValue
+            } else {
+                userCourseView?.isHidden = true
+                super.showsUserLocation = newValue
+            }
+        }
+    }
+    
+    var userLocationForCourseTracking: CLLocation?
+    var animatesUserLocation: Bool = false
+    
+    fileprivate let inactivityInterval: TimeInterval = 10
+    fileprivate let decreasedFrameInterval: Int = 12
+    
+    func resetInactivityTimer() {
+        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(inactivityTimerFinished), object: nil)
+        self.perform(#selector(inactivityTimerFinished), with: nil, afterDelay: inactivityInterval)
```

Comment:
I'm not seeing this fire currently. After moving the map away from followWithCourse, it just remains.

Reply:
Any gesture (rotate/pan/pinch) should trigger this. Could it be that fling (pan with velocity) doesn't trigger it?

Reply:
Oh, I was thinking this was a recenter timer after no activity.

Reply:
I do not see any call to `makeGestureRecognizersResetInactivityTimer` in this PR, is it the right behaviour ?

Reply:
Yes, throttling is currently not enabled.